<a href="https://colab.research.google.com/github/abhi-11nav/Airline-delay-prediction/blob/main/Airline_delay_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Importing the libraries for dataset import and manipulation

import pandas as pd 
import numpy as np 

In [2]:
# link for raw data uploaded on github

dataset_path = "https://raw.githubusercontent.com/abhi-11nav/Airline-delay-prediction/main/Airlines.csv"

In [3]:
# Loading the dataset into the dataset variable

dataset = pd.read_csv(dataset_path)

In [4]:
# Printing the first 5 rows of the dataset

dataset.head()

id Airline  Flight AirportFrom AirportTo  DayOfWeek  Time  Length  Delay
0   1      CO     269         SFO       IAH          3    15     205      1
1   2      US    1558         PHX       CLT          3    15     222      1
2   3      AA    2400         LAX       DFW          3    20     165      1
3   4      AA    2466         SFO       DFW          3    20     195      1
4   5      AS     108         ANC       SEA          3    30     202      0

In [5]:
# Printing the shape of data

dataset.shape

(539383, 9)

# DATA PREPROCESSING 

In [6]:
# Checking for missing values in the dataset 

missing_val_feature_count = 0

for col in dataset.columns:
  if dataset[col].isna().any() == True:
    missing_val_feature_count += 1

if missing_val_feature_count>0:
  print("WARNING - There are missing values in the dataset")
else:
  print("There are no missing values in the dataset")

There are no missing values in the dataset


This shows us that there are no missing values in the dataset

In [7]:
# Removing unnecessary feature "id" because it consists of all unique values

dataset.drop("id", axis=1,inplace=True)

"id" feature has been removed from the dataset

axis = 1 is to specify the axis to search on for removal of the feature,

inplace = True for removing the columns permanently from the data

.

**CLASSIFYING FEATURES AND LABELS**

DELAY IS OUR DEPENDENT VARIABLE (LABEL/Y)

REAMINING FEATURES ARE OUR INDEPENDENT VARIABLES ( FEATURES/X)

In [8]:
# Let us focus on the categorical variables of the dataset

# Printing the number of categorical variables
categorical_features = []

for col in dataset.columns:
  if type(dataset[col][0]) == str:
    categorical_features.append(col)

print(categorical_features,"are the categorical features in the dataset")

['Airline', 'AirportFrom', 'AirportTo'] are the categorical features in the dataset


In [9]:
# Now that we know how many categorical features are there in the dataset. Let us extract more information about it. 

unique_value_count = {}

for columns in categorical_features:
  unique_value_count[columns] = len(dataset[columns].value_counts())

print(unique_value_count)

{'Airline': 18, 'AirportFrom': 293, 'AirportTo': 293}


Airline feature has 18 unique values which is fine but the other two features "AirportTo" and "AirportFrom" has 293 unique values which have to be dealth with. 

In [13]:
"""# Encoding categorical variables  

#Creating columns with blank values to fill it with encoded values
for cols in categorical_features:
  dataset[cols+str("_encoded")] = 0 

for cols in categorical_features:
  for value in dataset[cols].unique():
    len = dataset[dataset[cols]==value]["Delay"].values.size
    label_ones = np.sum(dataset[dataset[cols]==value]["Delay"].values) 
    dataset[dataset[cols]==value][cols+str("_encoded")].replace((label_ones/(len-label_ones)))
    """

'# Encoding categorical variables  \n\n#Creating columns with blank values to fill it with encoded values\nfor cols in categorical_features:\n  dataset[cols+str("_encoded")] = 0 \n\nfor cols in categorical_features:\n  for value in dataset[cols].unique():\n    len = dataset[dataset[cols]==value]["Delay"].values.size\n    label_ones = np.sum(dataset[dataset[cols]==value]["Delay"].values) \n    dataset[dataset[cols]==value][cols+str("_encoded")].replace((label_ones/(len-label_ones)))\n    '

In [14]:
dataset.head()

Airline  Flight AirportFrom AirportTo  DayOfWeek  Time  Length  Delay  \
0      CO     269         SFO       IAH          3    15     205      1   
1      US    1558         PHX       CLT          3    15     222      1   
2      AA    2400         LAX       DFW          3    20     165      1   
3      AA    2466         SFO       DFW          3    20     195      1   
4      AS     108         ANC       SEA          3    30     202      0   

   Airline_encoded  AirportFrom_encoded  AirportTo_encoded  
0                0                    0                  0  
1                0                    0                  0  
2                0                    0                  0  
3                0                    0                  0  
4                0                    0                  0

In [16]:
dataset.drop(["Airline_encoded","AirportFrom_encoded","AirportTo_encoded"],axis=1,inplace=True)